superenv: fix calls to -R

jonchang opened the pull request:
`cc -R` is not the right way to pass rpaths to the linker. This pull request fixes that by modifying superenv. 

This might fix on Linux these issues: 

https://github.com/Homebrew/homebrew-science/pull/5065
https://github.com/h5py/h5py/issues/845

jonchang committed:
superenv: fix calls to -R

sjackman commented:
Thanks for this PR, Jon. It's helpful. I do hope to eventually merge `shims/linux/super/cc` and `shims/super/cc`.

sjackman commented:
> ```ruby
> path = canonical_path(val)
> ```

On Linuxbrew, we've been setting the `RPATH` to
```
~/.linuxbrew/opt/xz/lib
```
rather than
```
~/.linuxbrew/Cellar/xz/5.2.3/lib
```
so I'd be inclined to remove this line.
I had thought of possibly setting it to
```
~/.linuxbrew/opt/xz/lib:~/.linuxbrew/Cellar/xz/5.2.3/lib
```
 What are your thoughts?

sjackman commented:
There's two quite different use for RPATH though. One is so that a keg kind find its own libraries that are not stored in the usual `lib` directory, in which case `canonical_path` makes sense. The other is so that it could find the libraries of its dependencies, in which case it makes more sense to use the `opt_lib` symlink.

sjackman commented on Library/Homebrew/shims/linux/super/cc:
```diff
@@ -184,6 +184,11 @@ class Cmd
+      # -R is not the correct way to pass rpaths to the linker
+      val  = chuzzle($1) || enum.next
+      path = canonical_path(val)
+      args << "-Wl,-R," if keep?(path) && @lset.add?(path)
```
The `-R` option is non-standard, it should be `-Wl,-rpath=`.
See https://sourceware.org/binutils/docs/ld/Options.html#Options
> -R filename
--just-symbols=filename
Read symbol names and their addresses from filename, but do not relocate it or include it in the output. This allows your output file to refer symbolically to absolute locations of memory defined in other programs. You may use this option more than once.
For compatibility with other ELF linkers, if the -R option is followed by a directory name, rather than a file name, it is treated as the -rpath option.

sjackman replied:
> `&& @lset.add?(path)`

This doesn't seem right. If it's already in `lset` because it was added by `-L`, then the `-rpath` option won't be added as expected.

jonchang replied:
Thanks, I'll try this out. I couldn't find any documentation of either `-R` or `-rpath` but it makes sense that it's a linker flag rather than a `gcc` flag.

jonchang commented:
> `path = canonical_path(val)`
>
> There's two quite different use for RPATH though. One is so that a keg kind find its own libraries that are not stored in the usual lib directory, in which case canonical_path makes sense. The other is so that it could find the libraries of its dependencies, in which case it makes more sense to use the opt_lib symlink.

I'm OK with removing this line, as I just copied it from the above blocks. Though I think that we should punt the `opt` vs `cellar` directory to upstream, since I think that is something they are working on right now. It might be worth filing an issue about this at some point.

jonchang committed:
superenv: use standard linker flags

sjackman commented on Library/Homebrew/shims/linux/super/cc:
```diff
@@ -184,6 +184,11 @@ class Cmd
+      # -R is not the correct way to pass rpaths to the linker
+      val  = chuzzle($1) || enum.next
+      path = canonical_path(val)
+      args << "-Wl,-rpath=" if keep?(path) && @lset.add?(path)
```
> `&& @lset.add?(path)`

This doesn't seem right. If it's already in `lset` because it was added by `-L`, then the `-rpath` option won't be added as expected. How about using `@rset` instead? Add
```ruby
@rset = Set.new(rpath_paths)
```

jonchang replied:
Yeah, you're right about the `lset`. I've thought about this a bit more now and I think that this should just be an dumb translation of `-R` into `-Wl,-rpath=`, which will preserve current behavior. I'm thinking of also dropping the `keep?(path)` bit since it's possible that this could change certain behavior. `gcc` is often an undeclared runtime dependency of formulae built with gcc; and I think filtering paths here would change how formulae are currently built.

sjackman replied:
I'm okay with `keep?(path)`. It's useful to prevent adding wacky host paths like `/usr/lib64` to the `RPATH`.

> gcc is often an undeclared runtime dependency of formulae built with gcc; and I think filtering paths here would change how formulae are currently built.

Currently we add neither `glibc` nor `gcc` to the `RPATH`, meaning that those two formulae have to be linked, or everything breaks. That's a separate issue though. Quite relevant though to the upgrade path from `gcc-5` to `gcc-6`, which may be tricky.

sjackman replied:
`-Wl,` needs to be switched by `mode`.
```ruby
    wl = "-Wl," unless mode == :ld
    args << "#{wl}-rpath=" if keep?(path) && @rset.add?(path)
```

jonchang committed:
superenv: address review comment

- less aggressive filtering
- don't canonicalize path
- edit only on calls to `ld`

jonchang committed:
superenv: fix

sjackman commented:
Looks good to me, Jon. Thanks for this! Please squash this PR before merging, or use the GitHub `Squash and merge` button.

jonchang commented:
Thanks for the comments!

sjackman commented:
For some future PR if it's found to be needed, we ought to consider treating `-rpath` and `-Wl,-rpath=` similarly to `-L` and filter it through `if keep?(path)`, as this PR does for `-R`.
